File: fabric-samples/asset-transfer-secured-agreement/application-gateway-typescript/src/connect.ts
/*
 * Copyright IBM Corp. All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import * as grpc from '@grpc/grpc-js';
import { Identity, Signer, signers } from '@hyperledger/fabric-gateway';
import * as crypto from 'crypto';
import { promises as fs } from 'fs';
import * as path from 'path';

// MSP Id's of Organizations
export const mspIdRbi = 'RbiMSP';
export const mspIdHdfc = 'HdfcMSP';

// Path to Rbi crypto materials.
export const cryptoPathRbi = path.resolve(__dirname, '..', '..', '..', 'test-network', 'organizations', 'peerOrganizations', 'Rbi.example.com');

// Path to user private key directory.
export const keyDirectoryPathRbi = path.resolve(cryptoPathRbi, 'users', 'User1@Rbi.example.com', 'msp', 'keystore');

// Path to user certificate.
export const certDirectoryPathRbi = path.resolve(cryptoPathRbi, 'users', 'User1@Rbi.example.com', 'msp', 'signcerts');

// Path to peer tls certificate.
export const tlsCertPathRbi = path.resolve(cryptoPathRbi, 'peers', 'peer0.Rbi.example.com', 'tls', 'ca.crt');

// Path to Hdfc crypto materials.
export const cryptoPathHdfc = path.resolve(
    __dirname,
    '..',
    '..',
    '..',
    'test-network',
    'organizations',
    'peerOrganizations',
    'Hdfc.example.com'
);

// Path to Hdfc user private key directory.
export const keyDirectoryPathHdfc = path.resolve(
    cryptoPathHdfc,
    'users',
    'User1@Hdfc.example.com',
    'msp',
    'keystore'
);

// Path to Hdfc user certificate.
export const certDirectoryPathHdfc = path.resolve(
    cryptoPathHdfc,
    'users',
    'User1@Hdfc.example.com',
    'msp',
    'signcerts'
);

// Path to Hdfc peer tls certificate.
export const tlsCertPathHdfc = path.resolve(
    cryptoPathHdfc,
    'peers',
    'peer0.Hdfc.example.com',
    'tls',
    'ca.crt'
);
// Gateway peer endpoint.
export const peerEndpointRbi = 'localhost:7051';
export const peerEndpointHdfc = 'localhost:9051';

// Gateway peer container name.
export const peerNameRbi = 'peer0.Rbi.example.com';
export const peerNameHdfc = 'peer0.Hdfc.example.com';

// Collection Names
export const RbiPrivateCollectionName = 'RbiMSPPrivateCollection';
export const HdfcPrivateCollectionName = 'HdfcMSPPrivateCollection';

export async function newGrpcConnection(
    tlsCertPath: string,
    peerEndpoint: string,
    peerName: string
): Promise<grpc.Client> {
    const tlsRootCert = await fs.readFile(tlsCertPath);
    const tlsCredentials = grpc.credentials.createSsl(tlsRootCert);
    return new grpc.Client(peerEndpoint, tlsCredentials, {
        'grpc.ssl_target_name_override': peerName,
    });
}

export async function newIdentity(certDirectoryPath: string, mspId: string): Promise<Identity> {
    const certPath = await getFirstDirFileName(certDirectoryPath);
    const credentials = await fs.readFile(certPath);
    return { mspId, credentials };
}

export async function newSigner(keyDirectoryPath: string): Promise<Signer> {
    const keyPath = await getFirstDirFileName(keyDirectoryPath);
    const privateKeyPem = await fs.readFile(keyPath);
    const privateKey = crypto.createPrivateKey(privateKeyPem);
    return signers.newPrivateKeySigner(privateKey);
}

async function getFirstDirFileName(dirPath: string): Promise<string> {
    const files = await fs.readdir(dirPath);
    const file = files[0];
    if (!file) {
        throw new Error(`No files in directory: ${dirPath}`);
    }
    return path.join(dirPath, file);
}

File: fabric-samples/asset-transfer-private-data/application-gateway-go/connect.go
/*
Copyright 2024 IBM All Rights Reserved.

SPDX-License-Identifier: Apache-2.0
*/

package main

import (
	"crypto/x509"
	"fmt"
	"os"
	"path"

	"github.com/hyperledger/fabric-gateway/pkg/identity"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials"
)

const (
	cryptoPathRbi         = "../../test-network/organizations/peerOrganizations/Rbi.example.com"
	keyDirectoryPathRbi   = cryptoPathRbi + "/users/User1@Rbi.example.com/msp/keystore"
	certDirectoryPathRbi  = cryptoPathRbi + "/users/User1@Rbi.example.com/msp/signcerts"
	tlsCertPathRbi        = cryptoPathRbi + "/peers/peer0.Rbi.example.com/tls/ca.crt"
	peerEndpointRbi       = "dns:///localhost:7051"
	peerNameRbi           = "peer0.Rbi.example.com"
	cryptoPathHdfc        = "../../test-network/organizations/peerOrganizations/Hdfc.example.com"
	keyDirectoryPathHdfc  = cryptoPathHdfc + "/users/User1@Hdfc.example.com/msp/keystore"
	certDirectoryPathHdfc = cryptoPathHdfc + "/users/User1@Hdfc.example.com/msp/signcerts"
	tlsCertPathHdfc       = cryptoPathHdfc + "/peers/peer0.Hdfc.example.com/tls/ca.crt"
	peerEndpointHdfc      = "dns:///localhost:9051"
	peerNameHdfc          = "peer0.Hdfc.example.com"
)

// newGrpcConnection creates a gRPC connection to the Gateway server.
func newGrpcConnection(tlsCertPath, peerEndpoint, peerName string) *grpc.ClientConn {
	certificatePEM, err := os.ReadFile(tlsCertPath)
	if err != nil {
		panic(fmt.Errorf("failed to read TLS certificate file: %w", err))
	}

	certificate, err := identity.CertificateFromPEM(certificatePEM)
	if err != nil {
		panic(err)
	}

	certPool := x509.NewCertPool()
	certPool.AddCert(certificate)
	transportCredentials := credentials.NewClientTLSFromCert(certPool, peerName)

	connection, err := grpc.NewClient(peerEndpoint, grpc.WithTransportCredentials(transportCredentials))
	if err != nil {
		panic(fmt.Errorf("failed to create gRPC connection: %w", err))
	}

	return connection
}

// newIdentity creates a client identity for this Gateway connection using an X.509 certificate.
func newIdentity(certDirectoryPath, mspId string) *identity.X509Identity {
	certificatePEM, err := readFirstFile(certDirectoryPath)
	if err != nil {
		panic(fmt.Errorf("failed to read certificate file: %w", err))
	}

	certificate, err := identity.CertificateFromPEM(certificatePEM)
	if err != nil {
		panic(err)
	}

	id, err := identity.NewX509Identity(mspId, certificate)
	if err != nil {
		panic(err)
	}

	return id
}

// newSign creates a function that generates a digital signature from a message digest using a private key.
func newSign(keyDirectoryPath string) identity.Sign {
	privateKeyPEM, err := readFirstFile(keyDirectoryPath)
	if err != nil {
		panic(fmt.Errorf("failed to read private key file: %w", err))
	}

	privateKey, err := identity.PrivateKeyFromPEM(privateKeyPEM)
	if err != nil {
		panic(err)
	}

	sign, err := identity.NewPrivateKeySign(privateKey)
	if err != nil {
		panic(err)
	}

	return sign
}

func readFirstFile(dirPath string) ([]byte, error) {
	dir, err := os.Open(dirPath)
	if err != nil {
		return nil, err
	}

	fileNames, err := dir.Readdirnames(1)
	if err != nil {
		return nil, err
	}

	return os.ReadFile(path.Join(dirPath, fileNames[0]))
}
